FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,124 @@
+# Proposal: Xarray flexible indexes refactoring
+
+Current status: https://github.com/pydata/xarray/projects/1
+
+TODO: check if the project is up-to-date and update it if necessary.
+
+## 1. Data Model
+
+Indexes are used in Xarray to extract data from Xarray objects using coordinate labels instead of using integer array indices. Although the indexes used in an Xarray object can be accessed (or built on-the-fly) via public methods like `to_index()` or properties like `indexes`, those are mainly used internally.
+
+The goal of this project is to make those indexes 1st-class citizens of Xarray's data model. As such, indexes should clearly be separated from Xarray coordinates with the following relationships:
+
+- Index -> Coordinate: one-to-many
+- Coordinate -> Index: one-to-zero-or-one
+
+An index may be built from one or more coordinates. However, each coordinate must relate to one index at most. Additionally, a coordinate may not be tied to any index.
+
+The order in which multiple coordinates relate to an index should matter. For example, Scikit-Learn's `BallTree` index with the Haversine metric requires providing latitude and longitude values in that specific order. As another example, the order in which levels are defined in a `pandas.MultiIndex` may affect its lexsort depth.
+
+Xarray's current data model is already based on the same index-coordinate relationships. The current data model also implies a one-to-one relationship between a dimension and an index. This one-to-one relationship works as currently it is not possible to perform label-based data selection in Xarray using multi-dimensional coordinates, but since we want to enable this feature in this proposal, we need the dimension-index relationship to evolve towards many-to-many.
+
+In the example below, we'd like to select data points based on their x, y values and/or on their latitude/longitude positions:
+
+```
+>>> da
+<xarray.DataArray (x: 2, y: 2)>
+array([[5.4, 7.8],
+       [6.2, 4.7]])
+Coordinates:
+  * lon      (x, y) float64 10.2 15.2 12.6 17.6
+  * lat      (x, y) float64 40.2 45.6 42.2 47.6
+  * x        (x) float64 200.0 400.0
+  * y        (y) float64 800.0 1e+03
+
+>>> da.sel(x=..., y=..., lon=..., lat=...)
```

Comment:
This may not be the best example, because I suspect this particular case should be an error because lat/lon overlap with x/y. The desired behavior is ambiguous.

E.g., `da.sel(lon=..., lat=...)` and `ds.sel(x=..., y=...)` each should be fine but not the combination.

Reply:
Yes indeed. I was wondering how we should handle this:

- A. do not allow any combination
- B. add a new option in `.sel()` to specify how the query results from each index should be combined (i.e., union vs. intersection, etc.)